crypto/bsslmlkem/mlkem768.c

```diff
@@ -128,23 +134,59 @@ static int mlkem_init(void)
         || (encoded_public_key_size(RANK1024) != OSSL_MLKEM1024_PUBLIC_KEY_BYTES)
         || (ciphertext_size(RANK768) != OSSL_MLKEM768_CIPHERTEXT_BYTES)
         || (ciphertext_size(RANK1024) != OSSL_MLKEM1024_CIPHERTEXT_BYTES))
-        return 0;
+        goto err;
+
+    nctx->shake128_cache = EVP_MD_fetch(libctx, "SHAKE128", properties);
+    nctx->shake256_cache = EVP_MD_fetch(libctx, "SHAKE256", properties);
+    nctx->sha3_256_cache = EVP_MD_fetch(libctx, "SHA3-256", properties);
+    nctx->sha3_512_cache = EVP_MD_fetch(libctx, "SHA3-512", properties);
+    nctx->libctx = libctx;
+    if (properties != NULL)
+        nctx->properties = OPENSSL_strdup(properties);
+    if (nctx->shake128_cache == NULL || nctx->shake256_cache == NULL ||
+        nctx->sha3_256_cache == NULL || nctx->sha3_512_cache == NULL)
+        goto err;
+    return nctx;
+
+err:
+    ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+    ossl_mlkem_ctx_free(nctx);
+    return NULL;
+}
 
-    if (shake128_cache == NULL || shake256_cache == NULL ||
-        sha3_256_cache == NULL || sha3_512_cache == NULL) {
-        shake128_cache = EVP_MD_fetch(NULL, "SHAKE128", NULL);
-        shake256_cache = EVP_MD_fetch(NULL, "SHAKE256", NULL);
-        sha3_256_cache = EVP_MD_fetch(NULL, "SHA3-256", NULL);
-        sha3_512_cache = EVP_MD_fetch(NULL, "SHA3-512", NULL);
-        if (shake128_cache == NULL || shake256_cache == NULL ||
-            sha3_256_cache == NULL || sha3_512_cache == NULL) {
-            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
-            return 0;
-        }
+void ossl_mlkem_ctx_free(ossl_mlkem_ctx *ctx)
+{
+    if (ctx != NULL) {
+        if (ctx->shake128_cache != NULL)
+            EVP_MD_free(ctx->shake128_cache);
+        if (ctx->shake256_cache != NULL)
+            EVP_MD_free(ctx->shake256_cache);
+        if (ctx->sha3_256_cache != NULL)
+            EVP_MD_free(ctx->sha3_256_cache);
+        if (ctx->sha3_512_cache != NULL)
+            EVP_MD_free(ctx->sha3_512_cache);
+        OPENSSL_free(ctx->properties);
     }
-    return 1;
+    OPENSSL_free(ctx);
 }
 
+/*
+ * TODO(ML-KEM): ctx validation: Is this over the top checking an input
+ * construction of which has been checked before? But a malign user
+ * could cause problems changing the struct before passing it back
+ * Problem: This costs a discernible level of performance (~3%)
+ */
+static int validate_mlkem_ctx(ossl_mlkem_ctx *ctx)
```

Comment:
IMO this is unnecessary although I would be very much surprised if the performance impact was really discernible.

For internal APIs there is no concern about malign users and the check in ossl_mlkem_newctx() should be fully sufficient.

Reply:
> IMO this is unnecessary although I would be very much surprised if the performance impact was really discernible.

I regularly run `speed` and see visible differences with this present (2-3%), so would be happy to drop this from that perspective.

> For internal APIs there is no concern about malign users and the check in ossl_mlkem_newctx() should be fully sufficient.

OK with me -- if this is/stays internal: Is the definition for that presence/absence in the global symbol table?

Reply:
removed in latest commit.